File: problem37/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans = 0;
    public int longestZigZag(TreeNode root) {
        zigzag(root.left, -1);
        zigzag(root.right, 1);
        return ans;
    }
    private int zigzag(TreeNode root, int branch) {
        if (root == null)
            return 0;
        int tmp = 1;
        int left = zigzag(root.left, -1);
        int right = zigzag(root.right, 1);
        if (branch == -1)
            tmp += right;
        else
            tmp += left;
        ans = Math.max(ans, tmp);
        return tmp;
    }
}

File: problem37/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        longest = 0
        def zigzag_dfs(root, dir):
            if not root: return 0
            nonlocal longest
            curr_longest = 1

            left = zigzag_dfs(root.left, 'l')
            right = zigzag_dfs(root.right, 'r')

            if dir == 'l': curr_longest += right
            else: curr_longest += left
            
            longest = max(longest, curr_longest)
            return curr_longest

        zigzag_dfs(root.left, 'l')
        zigzag_dfs(root.right, 'r')
        return longest
